src/oss-reproducible/Helpers/Helpers.cs

```diff
@@ -0,0 +1,221 @@
+                !Directory.Exists(leftDirectory) || !Directory.Exists(rightDirectory))
+            {
+                Logger.Warn("Directory does not exist.");
+                throw new DirectoryNotFoundException("Directory does not exist.");
+            }
+
+            var leftContent = GenerateDirectoryHashes2(packageUrl, strategyName, leftDirectory);
+            var rightContent = GenerateDirectoryHashes2(packageUrl, strategyName, rightDirectory);
+            results.AddRange(rightContent.Keys.Except(leftContent.Keys).Select(f => new DirectoryDifference { Filename = string.Join(',', rightContent[f]), Operation = DirectoryDifferenceOperation.Added }));
+            results.AddRange(leftContent.Keys.Except(rightContent.Keys).Select(f => new DirectoryDifference { Filename = string.Join(',', leftContent[f]), Operation = DirectoryDifferenceOperation.Removed }));
+
+            return results;
+        }
+
+        /*
+        internal static IEnumerable<KeyValuePair<string, string>> GenerateDirectoryHashes(PackageURL packageUrl, string directory, bool useFilters = true)
+        {
+            var hashFunction = SHA256.Create();
+            var directoryFiles = Directory.EnumerateFiles(directory, "*", new EnumerationOptions { RecurseSubdirectories = true, AttributesToSkip = 0 });
+            var filteredFiles = directoryFiles.Where(f => !IgnoreFilter.IsIgnored(packageUrl, this.GetType().Name, f));
+            foreach (var file in filteredFiles)
+            {
+                var hashValue = hashFunction.ComputeHash(File.ReadAllBytes(file));
+                var hashString = Convert.ToHexString(hashValue);
+                yield return KeyValuePair.Create<string, string>(file, hashString);
+            }
+        }
+        */
+
+        internal static Dictionary<string, HashSet<string>> GenerateDirectoryHashes2(PackageURL packageUrl, string strategyName, string directory)
+        {
+            var results = new Dictionary<string, HashSet<string>>();
+            var hashFunction = SHA256.Create();
+            var directoryFiles = Directory.EnumerateFiles(directory, "*", new EnumerationOptions { RecurseSubdirectories = true, AttributesToSkip = 0 });
+            var filteredFiles = directoryFiles.Where(f => !IgnoreFilter.IsIgnored(packageUrl, strategyName, f));
+            foreach (var filename in filteredFiles)
+            {
+
+                var hashValue = hashFunction.ComputeHash(File.ReadAllBytes(filename));
+                var hashString = Convert.ToHexString(hashValue);
+                if (results.ContainsKey(hashString))
+                {
+                    results[hashString].Add(filename);
+                }
+                else
+                {
+                    results[hashString] = new HashSet<string>() { filename };
+                }
+            }
+            return results;
+        }
+
+
+        internal static string? GetFirstNonSingularDirectory(string? directory)
```

Comment:
It appears to:

If the target directory does not exist:
return null
If the target directory has exactly one entry in it and that is a directory
recurse on it
In all other cases
return the directory

Reply:
Yep, since RecursiveExtractor tends to create a couple parent directories, we want to ignore them. We basically want to get to the first interesting directory.